private/buf/bufgen/generator.go

```diff
@@ -239,3 +509,53 @@ type generateOptions struct {
 func newGenerateOptions() *generateOptions {
 	return &generateOptions{}
 }
+
+type localPluginResponse struct {
+	responseWriter appproto.ResponseWriter
+	index          int
+}
+
+func newLocalPluginResponse(responseWriter appproto.ResponseWriter, index int) *localPluginResponse {
+	return &localPluginResponse{
+		responseWriter: responseWriter,
+		index:          index,
+	}
+}
+
+type remotePlugin struct {
+	reference *registryv1alpha1.PluginReference
+	index     int
+}
+
+func newRemotePlugin(reference *registryv1alpha1.PluginReference, index int) *remotePlugin {
+	return &remotePlugin{
+		reference: reference,
+		index:     index,
+	}
+}
+
+type semaphore chan struct{}
+
+func newSemaphore(limit int) semaphore {
+	return make(semaphore, limit)
+}
+
+func (s semaphore) Acquire(ctx context.Context) error {
+	select {
+	case s <- struct{}{}:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (s semaphore) Release(ctx context.Context) {
+	select {
+	case <-s:
+	case <-ctx.Done():
+	}
+}
+
+func (s semaphore) Discard() {
+	close(s)
+}
```

Comment:
Curious, why not https://pkg.go.dev/golang.org/x/sync/semaphore? Is the thought that this is simple enough as-is?

Reply:
Yes, this is simple enough, no need for weights. It's just syntactic sugar around a channel.